lib/kastlex/token_storage.ex

```diff
@@ -0,0 +1,153 @@
+    GenServer.start_link(__MODULE__, options, name: @server)
+  end
+
+  def register(sub, token_hash, exp) do
+    GenServer.call(@server, {:register, sub, token_hash, exp})
+  end
+
+  def revoke(sub) do
+    GenServer.call(@server, {:revoke, sub})
+  end
+
+  def init(options) do
+    :ets.new(@tokens_table, [:set, :protected, :named_table, {:read_concurrency, true}])
+    client = options.brod_client_id
+    :ok = :brod.start_consumer(client, @topic, [])
+    :ok = :brod.start_producer(client, @topic, [])
+    send self, :subscribe
+    {:ok, %{:client => client}}
+  end
+
+  def handle_call({:register, sub, token_hash, exp}, _from, state) do
+    data = :erlang.term_to_binary(%{:token_hash => token_hash, :exp => exp})
+    case :brod.produce_sync(state.client, @topic, 0, sub, data) do
+      :ok ->
+        :ets.insert(@tokens_table, {sub, data})
+        Logger.debug "successfully registered token for #{sub}"
+        {:reply, :ok, state}
+      {:error, _} = error ->
+        Logger.error "failed to send token to kafka: #{inspect error}"
+        {:reply, error, state}
+    end
+  end
+
+  def handle_call({:revoke, sub}, _from, state) do
+    case :brod.produce_sync(state.client, @topic, 0, sub, "") do
+      :ok ->
+        :ets.delete(@tokens_table, sub)
+        Logger.debug "successfully deleted token for #{sub}"
+        {:reply, :ok, state}
+      {:error, _} = error ->
+        Logger.error "failed to delete token from kafka: #{inspect error}"
+        {:reply, error, state}
+    end
+  end
+
+  def handle_info(:subscribe, state) do
+    options = [{:begin_offset, :earliest},
+               {:prefetch_count, 1_000}
+              ]
+    case :brod.subscribe(state.client, self(), @topic, 0, options) do
+      {:ok, pid} ->
+        Logger.debug "subscribed on #{@topic}"
+        _ = Process.monitor(pid)
+        {:noreply, Map.put(state, :consumer, pid)}
+      {:error, _reason} ->
+        # consumer down, client down, etc. try again later
+        resubscribe()
+    end
+  end
+
+  def handle_info({pid, msg_set}, %{consumer: pid} = state)
+  when Record.is_record(msg_set, :kafka_message_set) do
+    messages = kafka_message_set(msg_set, :messages)
+    last_offset = kafka_message(List.last(messages), :offset)
+    Enum.each(messages, fn(msg) -> handle_message(msg) end)
+    :ok = :brod.consume_ack(pid, last_offset)
+    {:noreply, state}
+  end
+
+  def handle_info({pid, fetch_error}, %{consumer: pid} = state)
+  when Record.is_record(fetch_error, :kafka_fetch_error) do
+    Logger.info "fetch error: #{inspect fetch_error}"
+    resubscribe()
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, _ref, :process, pid, reason}, %{consumer: pid} = state) do
+    Logger.info "consumer process died: #{inspect reason}"
+    resubscribe()
+    {:noreply, state}
+  end
+
+  defp handle_message(msg) do
+    key = kafka_message(msg, :key)
+    value = kafka_message(msg, :value)
+    case value do
+      :undefined ->
```

Comment:
match `<<>>` too.
just in case we change kafka_protodocl behaviour in future releases.